Improve the MSBuild targets documentation

nkolev92 opened the pull request:
Fixes https://github.com/NuGet/docs.microsoft.com-nuget/issues/1061 https://github.com/NuGet/docs.microsoft.com-nuget/issues/1358
https://github.com/NuGet/Home/issues/7415


Supersedes https://github.com/NuGet/docs.microsoft.com-nuget/pull/1102

Document the missing properties, add a note about msbuild treating env vars as properties and document `-restore`

New issue discovered during this doc writing: https://github.com/NuGet/docs.microsoft.com-nuget/issues/1394

karann-msft commented:
Docs Build status updates of commit _[a8ea0e7](https://github.com/NuGet/docs.microsoft.com-nuget/pull/1395/commits/a8ea0e779dd7ed093f89bf61ad6fd590fc442428)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/reference/msbuild-targets.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/reference/msbuild-targets.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/reference/msbuild-targets?branch=pr-en-us-1395) |
[docs/tools/cli-ref-environment-variables.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/tools/cli-ref-environment-variables.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/tools/cli-ref-environment-variables?branch=pr-en-us-1395) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C4%5C11%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201904110025466776-1395%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=Qq%2B3yFEusnZNQtIfOZMW653dW%2FudyBF5CuvPxUXNnTg%3D&st=2019-04-11T00%3A21%3A52Z&se=2019-05-12T00%3A26%3A52Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

karann-msft commented:
Docs Build status updates of commit _[7605ed3](https://github.com/NuGet/docs.microsoft.com-nuget/pull/1395/commits/7605ed3c963362b07a2d29f7d0d19667321cce01)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/tools/cli-ref-environment-variables.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/tools/cli-ref-environment-variables.md) | :bulb:Suggestion | [View](https://review.docs.microsoft.com/en-us/nuget/tools/cli-ref-environment-variables?branch=pr-en-us-1395) | [Details](#user-content-e714ca7c-d620c61b)
[docs/reference/msbuild-targets.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/reference/msbuild-targets.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/reference/msbuild-targets?branch=pr-en-us-1395) |

<a id="e714ca7c-d620c61b"></a>
### [docs/tools/cli-ref-environment-variables.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/tools/cli-ref-environment-variables.md)
 - **Line 34**: **[Suggestion]** Illegal link: `&lt;a href=&quot;#https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference&quot;&gt;properties&lt;/a&gt;` -- missing bookmark. The file tools/cli-ref-environment-variables.md doesn&#39;t contain a bookmark named https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference.

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C4%5C11%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201904110025575624-1395%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=V%2BIlThek2VERACA3ATRNNDOTSFow1ULsa8l76YpYjIA%3D&st=2019-04-11T00%3A23%3A09Z&se=2019-05-12T00%3A28%3A09Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

karann-msft commented:
Docs Build status updates of commit _[0ceefbb](https://github.com/NuGet/docs.microsoft.com-nuget/pull/1395/commits/0ceefbb12298ad0a9bfb96efaa908da07ea07ef0)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/tools/cli-ref-environment-variables.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/tools/cli-ref-environment-variables.md) | :bulb:Suggestion | [View](https://review.docs.microsoft.com/en-us/nuget/tools/cli-ref-environment-variables?branch=pr-en-us-1395) | [Details](#user-content-e714ca7c-248ce53a)
[docs/reference/msbuild-targets.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/reference/msbuild-targets.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/reference/msbuild-targets?branch=pr-en-us-1395) |

<a id="e714ca7c-248ce53a"></a>
### [docs/tools/cli-ref-environment-variables.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/tools/cli-ref-environment-variables.md)
 - **Line 16**: **[Suggestion]** Illegal link: `&lt;a href=&quot;#https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference&quot;&gt;properties&lt;/a&gt;` -- missing bookmark. The file tools/cli-ref-environment-variables.md doesn&#39;t contain a bookmark named https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference.

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C4%5C11%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201904110029578070-1395%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=mC4ggZQMfVzJA7VlnBhnyinXSGeu7DGIgaEoecACkCo%3D&st=2019-04-11T00%3A26%3A05Z&se=2019-05-12T00%3A31%3A05Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

karann-msft commented:
Docs Build status updates of commit _[b53290d](https://github.com/NuGet/docs.microsoft.com-nuget/pull/1395/commits/b53290d8c88f57351d9dbe07ab4d818b70f4754f)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/reference/msbuild-targets.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/reference/msbuild-targets.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/reference/msbuild-targets?branch=pr-en-us-1395) |
[docs/tools/cli-ref-environment-variables.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/tools/cli-ref-environment-variables.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/tools/cli-ref-environment-variables?branch=pr-en-us-1395) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C4%5C11%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201904110032469406-1395%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=7VETzFfFQkHfdys62pR3rrTN42MHUjiQo8OH5ONpTTY%3D&st=2019-04-11T00%3A28%3A57Z&se=2019-05-12T00%3A33%3A57Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

nkolev92 commented:
👋 

@zivkan @karann-msft @anangaur

karann-msft commented on docs/reference/msbuild-targets.md:
```diff
@@ -366,6 +371,23 @@ Restore creates the following files in the build `obj` folder:
 | `{projectName}.projectFileExtension.nuget.g.props` | References to MSBuild props contained in packages |
 | `{projectName}.projectFileExtension.nuget.g.targets` | References to MSBuild targets contained in packages |
 
+### Restore and build in the same MSBuild invocation
```
this section seems to come out of nowhere. Is there an error associated when the user does `msbuild -t:restore,build` ?

Or it should be written as, "Multiple operations in the same MSBuild invocation", then talk about some good examples, and then talk about some bad examples.

nkolev92 replied:
There are no specific error associated with the above, which is why it's such a big problem. 
You don't know it doesn't work until it's too late. 

It's not really multiple operations during the same msbuild invocation. 

The MSBUild side switch is explicitly added for the restore target. 
Only restore can be invoked with `msbuild -restore -t:SomeTargetName`

zivkan replied:
How about "Restoring and building with one MSBuild command"? Sure, technically the issue is with restoring and any other target in the same command, but I imagine that most people who hit this are building and this title is more likely to catch their eye.

nkolev92 replied:
I like that.

karann-msft replied:
let's run with Andy's suggestion.

zivkan commented on docs/reference/msbuild-targets.md:
```diff
@@ -336,9 +336,14 @@ Additional restore settings may come from MSBuild properties in the project file
 | RestoreNoCache | If true, avoids using cached packages. See [Managing the global packages and cache folders](../consume-packages/managing-the-global-packages-and-cache-folders.md). |
 | RestoreIgnoreFailedSources | If true, ignores failing or missing package sources. |
+| RestoreFallbackFolders | Fallback folders, used in the same way the user packages folder is used. |
+| RestoreAdditionalProjectSources | Additional sources to the resulting list of sources |
```
Additional sources to use during restore?

"resulting list" just sounds strange to me.

zivkan commented on docs/reference/msbuild-targets.md:
```diff
@@ -336,9 +336,14 @@ Additional restore settings may come from MSBuild properties in the project file
 | RestoreIgnoreFailedSources | If true, ignores failing or missing package sources. |
+| RestoreFallbackFolders | Fallback folders, used in the same way the user packages folder is used. |
+| RestoreAdditionalProjectSources | Additional sources to the resulting list of sources |
+| RestoreAdditionalProjectFallbackFolders | Additional fallback folders appended to the resulting list of fallback folders |
```
Additional fallback folders?
Additional fallback folders to use during restore?

zivkan commented on docs/reference/msbuild-targets.md:
```diff
@@ -336,9 +336,14 @@ Additional restore settings may come from MSBuild properties in the project file
 | RestoreTaskAssemblyFile | Path to `NuGet.Build.Tasks.dll`. |
 | RestoreGraphProjectInput | Semicolon-delimited list of projects to restore, which should contain absolute paths. |
-| RestoreOutputPath | Output folder, defaulting to the `obj` folder. |
+| RestoreUseSkipNonexistentTargets  | When the projects are collected via MSBuild it determines whether they collected using the `SkipNonexistentTargets` optimization. When not set, defaults to `true`. Consequence is fail-fast behavior when a project's targets cannot be imported. |
```
whether they **are** collected.

**The** consequence is **a** fail-fast behaviour

zivkan commented on docs/tools/cli-ref-environment-variables.md:
```diff
@@ -13,6 +13,9 @@ The behavior of the nuget.exe CLI can be configured through a number of environm
 
 In general, options specified directly on the command line or in NuGet configuration files have precedence, but there are a few exceptions such as *FORCE_NUGET_EXE_INTERACTIVE*. If you find that nuget.exe behaves differently between different computers, an environment variable could be the cause. For example, Azure Web Apps Kudu (used during deployment) has *NUGET_XMLDOC_MODE* set to *skip* to sp ...
 
+The NuGet CLI uses MSBuild to read the project files. All environment variables are available as [properties](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference) during the MSBuild evaluation.
```
This isn't the correct syntax for linking docs from other repos: https://review.docs.microsoft.com/en-us/help/contribute/links-how-to?branch=master#links-to-articles-in-other-repos-or-docsets

In particular, if someone's already browsing the docs in a language other than en-us, this link would take them to docs in the wrong language.

karann-msft commented:
Docs Build status updates of commit _[5287d85](https://github.com/NuGet/docs.microsoft.com-nuget/pull/1395/commits/5287d8514b0659a25cfb98e9e93be849ac5cfee5)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/reference/msbuild-targets.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/reference/msbuild-targets.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/reference/msbuild-targets?branch=pr-en-us-1395) |
[docs/tools/cli-ref-environment-variables.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/tools/cli-ref-environment-variables.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/tools/cli-ref-environment-variables?branch=pr-en-us-1395) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C4%5C17%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201904172303351720-1395%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=LhBkHtDBzhRG2WnZY65WtVdiAddTqZhyb4PiosrlRXY%3D&st=2019-04-17T22%3A59%3A42Z&se=2019-05-18T23%3A04%3A42Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

nkolev92 commented:
@zivkan Addressed your feedback. 

@karann-msft Happy to brainstorm other ways of wording that section. 

It's very much a restore specific thing.

karann-msft commented:
Docs Build status updates of commit _[38f04e4](https://github.com/NuGet/docs.microsoft.com-nuget/pull/1395/commits/38f04e4b72dfd52affbfa65144ea844f7e16c07d)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/reference/msbuild-targets.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/reference/msbuild-targets.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/reference/msbuild-targets?branch=pr-en-us-1395) |
[docs/tools/cli-ref-environment-variables.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/tools/cli-ref-environment-variables.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/tools/cli-ref-environment-variables?branch=pr-en-us-1395) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C4%5C17%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201904172304295557-1395%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=U2%2FMhLJc8kzhISzI5hbv9gVXR8xwly4ocKnhMa7lA44%3D&st=2019-04-17T23%3A00%3A58Z&se=2019-05-18T23%3A05%3A58Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

karann-msft commented:
Docs Build status updates of commit _[08839dd](https://github.com/NuGet/docs.microsoft.com-nuget/pull/1395/commits/08839dd56541028e30281a5118c60b1cfbd24ad1)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/reference/msbuild-targets.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/reference/msbuild-targets.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/reference/msbuild-targets?branch=pr-en-us-1395) |
[docs/tools/cli-ref-environment-variables.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master-nkolev92-msbuildProperties/docs/tools/cli-ref-environment-variables.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/tools/cli-ref-environment-variables?branch=pr-en-us-1395) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C4%5C18%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201904180033105813-1395%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=XCA6xOOY9BCb6D0ty08jGNZvhUMnjmlGh2Tgl6zJYX4%3D&st=2019-04-18T00%3A29%3A21Z&se=2019-05-19T00%3A34%3A21Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.
